src/common/src/types/struct_type.rs

```diff
@@ -29,112 +29,112 @@ pub struct StructType(Arc<StructTypeInner>);
 impl Debug for StructType {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("StructType")
-            .field("field_names", &self.0.field_names)
-            .field("field_types", &self.0.field_types)
+            .field("fields", &self.0.fields)
             .finish()
     }
 }
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
 struct StructTypeInner {
-    /// Details about a struct type. There are 2 cases for a struct:
-    /// 1. `field_names.len() == field_types.len()`: it represents a struct with named fields,
-    ///     e.g. `STRUCT<i INT, j VARCHAR>`.
-    /// 2. `field_names.len() == 0`: it represents a struct with unnamed fields,
-    ///     e.g. `ROW(1, 2)`.
-    field_names: Box<[String]>,
-    field_types: Box<[DataType]>,
+    /// The name and data type of each field.
+    ///
+    /// If fields are unnamed, the names will be `f1`, `f2`, etc.
+    fields: Box<[(String, DataType)]>,
+    /// Whether the fields are unnamed.
+    is_unnamed: bool,
 }
 
 impl StructType {
     /// Creates a struct type with named fields.
     pub fn new(named_fields: impl IntoIterator<Item = (impl Into<String>, DataType)>) -> Self {
-        let iter = named_fields.into_iter();
-        let mut field_types = Vec::with_capacity(iter.size_hint().0);
-        let mut field_names = Vec::with_capacity(iter.size_hint().0);
-        for (name, ty) in iter {
-            field_names.push(name.into());
-            field_types.push(ty);
-        }
+        let fields = named_fields
+            .into_iter()
+            .map(|(name, ty)| (name.into(), ty))
+            .collect();
+
         Self(Arc::new(StructTypeInner {
-            field_types: field_types.into(),
-            field_names: field_names.into(),
+            fields,
+            is_unnamed: false,
         }))
     }
 
-    /// Creates a struct type with no fields.
+    /// Creates a struct type with no fields. This makes no sense in practice.
     #[cfg(test)]
     pub fn empty() -> Self {
-        Self(Arc::new(StructTypeInner {
-            field_types: Box::new([]),
-            field_names: Box::new([]),
-        }))
+        Self::unnamed(Vec::new())
     }
 
-    /// Creates a struct type with unnamed fields.
+    /// Creates a struct type with unnamed fields. The names will be assigned `f1`, `f2`, etc.
     pub fn unnamed(fields: Vec<DataType>) -> Self {
+        let fields = fields
+            .into_iter()
+            .enumerate()
+            .map(|(i, ty)| (format!("f{}", i + 1), ty))
```

Comment:
We can simplify this part now:
https://github.com/risingwavelabs/risingwave/blob/7df444c14f5b78effff4e9634a9488befdcd6467/src/expr/impl/src/scalar/to_jsonb.rs#L253-L257